### ISSUE TITLE
Lasers acting weird

### ISSUE THREAD

**Issue by Aivech** [ReikaKalseki/Reika_Mods_Issues#1099]
Using v15A, in creative. Laser turrets are not doing any damage to mobs or terrain, except at much closer ranges than expected. The lasers are being powered by ElectriCraft motors.

Angular transducer readout:
Power is being received from: [None]
Range: 256m. Max range: 256m. 
Laser gun receiving 33.423 MW @ 8192 rad/s.

Observed range is less than 30 meters absolute distance according to Bibliocraft's tape measure.

Yes, I did read the handbook entry.

Additionally, turrets ignore line of sight when selecting a target.

**Comment by Aivech** [comment]
EDIT: Upon further investigation this appears to be severe desync...on a singleplayer world.
The beam is pointing at one location visually but it actually exists somewhere else.

**Comment by ReikaKalseki** [comment]
Angles _should_ be synced...

**Comment by Aivech** [comment]
On further observation:
1. On first turning the lasers on they properly select a target and point at it. The death ray is there and the laser does point towards the target, which is usually a slime.
2. Once this target inevitably dies, the laser's model and beam render point towards a new target, but the death ray remains pointed at the old target location. Occasionally, the model and beam "jump" back to where the death ray actually is, then sweep towards their target again.
3. The terrain outside is covered in radial beams of sand and glass as would be expected were they working properly, but I never see them do it.

I have tested the TPS and it is healthy.

TL;DR: Render retargets after killing mob, but actual fire beam does not. Render sometimes jumps back to where the fire beam is.

**Comment by ReikaKalseki** [comment]
WTF....

**Comment by ReikaKalseki** [comment]
Is this a dedicated server?

**Comment by Gaelan** [comment]
He said "in a singleplayer world" a few comments up.

**Comment by ReikaKalseki** [comment]
Then I have no idea.

**Comment by ReikaKalseki** [comment]
Do others have this issue?

**Comment by Barhandar** [comment]
Yes, the issue STILL persists, except now it doesn't desync. That is, not only deathbeam is stuck in place, but the visual of the turret, too.

But do note that I'm playing on dedicated server running on same machine, so integrated might still experience the desync out of nowhere.

**Comment by OvermindDL1** [comment]
Just from a cursory look at the code, both `TileEntityLaserGun` and `TileEntityAimedCannon` I'm not seeing where `target` is updated from a pack or other sync measure, it looks like it is only ever updated by calling `getTarget`, which is only ever called from:
https://github.com/ReikaKalseki/RotaryCraft/blob/b8203ebfc4727306cf300c56336050d2afa56389/Base/TileEntity/TileEntityAimedCannon.java#L126

Which is inside of `updateTileEntity`, which from searching all of Reika's Mods appears to only be called from:
https://github.com/ReikaKalseki/DragonAPI/blob/cf52c3e3466b84f4b9198a7e20c4794952f05291/Base/TileEntityBase.java#L509

And this `updateEntity` function is called from a short variety of places:
```java
ChromatiCraft/TileEntity/AOE/Effect/TileEntityAccelerator.java:                                 te.updateEntity();
ChromatiCraft/Block/BlockHeatLamp.java:                                 te.updateEntity();
DragonAPI/ModInteract/DeepInteract/TinkerSmelteryHandler.java:          te.updateEntity();
```
Including also base minecraft itself via World.java, so that should be ticked on the client side.

Since this happens to both lasers as well as other turret types such as the gatling (as per linked issue), then the issue likely resides in `TileEntityAimedCannon`, so looking further in it.

First thing to notice is that there is this code in the `updateTileEntity` function:
```java
		if (worldObj.isRemote)
			;//return;
```
So that's commented out, so the code is ran on both client and server (though considering Entity's can have different positions on both client and server, it seems like this could desync really really easy, although it would be graphical only).

Second thing noticed is that `isCustomAim` in `updateTileEntity` will ignore the `getTarget` call, so seeing if anything sets it to to true anywhere, only place it is ever set at all is when loading it from NBT:
https://github.com/ReikaKalseki/RotaryCraft/blob/b8203ebfc4727306cf300c56336050d2afa56389/Base/TileEntity/TileEntityAimedCannon.java#L223

However an NBT with tag of `"aim"` is not set anywhere except in the same NBT setting function, so it does not appear to ever be touched even by subclasses.  I'm guessing it's uncleaned dead code.

Third thing that could potentially cause it to not run is this code in `updateTileEntity`:
```java
		if (power < MINPOWER)
			return;
```
So where does `power` get set, not in this class, going back up the parents finds it defined in the grandparent of `TileEntityIOMachine`.  Following the usages of `power` through that class, the parent, and grandparents finds:

* It is initially set in `writeSyncTag`/`readSyncTag`.
* It is not touched otherwise in `TileEntityIOMachine`.
* It is set in `TileEntityPowerReceiver` at:  https://github.com/ReikaKalseki/RotaryCraft/blob/b8203ebfc4727306cf300c56336050d2afa56389/Base/TileEntity/TileEntityPowerReceiver.java#L277 if the machine is not `doubleSided`.  (At which point `prevpower` is updated to it if they don't match, nothing else is done, what's with all this unclean dead code?!)
* And it's updated again at https://github.com/ReikaKalseki/RotaryCraft/blob/b8203ebfc4727306cf300c56336050d2afa56389/Base/TileEntity/TileEntityPowerReceiver.java#L357 later in the same function, ignoring and never using the not `doubleSided` power set of it (while also uselessly updating `prevpower`).
* It's set to 0 at https://github.com/ReikaKalseki/RotaryCraft/blob/b8203ebfc4727306cf300c56336050d2afa56389/Base/TileEntity/TileEntityPowerReceiver.java#L448 if the other TE is a WorldRift (shouldn't be the issue I'd think though).
* And of course it's set yet again the same way as the above ones at https://github.com/ReikaKalseki/RotaryCraft/blob/b8203ebfc4727306cf300c56336050d2afa56389/Base/TileEntity/TileEntityPowerReceiver.java#L460 in `getOffsetPower4Sided` (so much duplicate code!).
* And yet again at https://github.com/ReikaKalseki/RotaryCraft/blob/b8203ebfc4727306cf300c56336050d2afa56389/Base/TileEntity/TileEntityPowerReceiver.java#L541 in `getSummativeSidedPower` (yeesh getters should never ever set data like is accessible externally of the function!).

The `MINPOWER` that it's compared to is set in the initializer of `TileEntityPowerReceiver`, which is set to 0 if `PowerReceivers.hasMultiValuedPower` is false (it is not in this case), or else it gets set to `PowerReceivers.  So in this case MINPOWER for, say, the laser turret, will be a value of 8388608 (so plenty high).

So considering this:
```java
		if (power < MINPOWER)
			return;
```
The power needs to be pretty high for it to run (over 8MW), but if it runs on server then client 'should' be fine, *if* that value is updated on the clientside.  So of where `power` is set, let's go through those:

* `getPower(World world, int x, int y, int z, boolean doubleSided)` in `TileEntityPowerReceiver` has a `if (worldObj.isRemote && !RotaryAux.getPowerOnClient) return;`, so what does `RotaryAux.getPowerOnClient` infer, it seems to be `public static final boolean getPowerOnClient = ConfigRegistry.POWERCLIENT.getState() || ReikaObfuscationHelper.isDeObfEnvironment();`.
* The other usages are gated the same as above.

So, `ConfigRegistery.POWERCLIENT` is set as:
```java
	POWERCLIENT("Run power transfer code on client", false),  //caused many issues
```
So it seems to default to `false`, but when it is `false` then the Turret's won't update their clientside target after initial sync.

So, now I have a question for @Barhandar or @DaMachinator or whoever, I need to see your configuration option for `Run power transfer code on client` and see if it is set to `false` (the default).  Try setting it to `true`, does that make turrets work on the client?

I'm unsure what these `//caused many issues` is referencing, let's check `git blame`, so Reika was the one that made that change in 4e51c7e051e2b13e2332f5447773ce5d99f3e8c5 and that was only uncommenting it (wtf?) and the one that commented it out was f0a6ae062a1a00e3a1070fec43baf88d37fc62fd and it was hardcoded to `true` at this point and so let's check the comments for why these changes were made... oh yeah, Reika doesn't use git properly, this commits are bloody massive with no documentation for 'why' changes were made at all...

So yeah, setting it to `true` should get it working, but no clue what other problems it might cause considering the extremely non-descriptive git commits and git comments (this whole repository makes my code-review senses not just tingle but scream with overwhelming pain...).

**Comment by Barhandar** [comment]
@OvermindDL1 as I've said, I am using a dedicated server and pretty sure targetted mob wouldn't stay alive if the turret was active on server, but not client. Nonetheless that probably unrelated (failure to move to the target/target dropping after few ticks isn't the same as losing clientside power) option is set to `false`.

And even originally it worked on client but not _on server_ anyway.

**Comment by OvermindDL1** [comment]
> And even originally it worked on client but not on server anyway.

Yeah there is odd stuff all around this code...  ^.^;

If only he used a real actual normal java build system then it would be pretty trivial to debug the live code.  I can with my copy but running the mods is... laggy at the best of times.
